## ML&Fintech HW7 Chapter5 Q9

Name: Alvin B. Lin\
ID: 112652040\
Date: 14/11/2025\
HW #: 7

9.We will now consider the `Boston` housing data set, from the `ISLP` library.

(a) Based on this data set, provide an estimate for the population mean of `medv`. Call this estimate $\hat{\mu}$.

In [5]:
import ISLP
import numpy as np
import pandas as pd

In [4]:
Boston = ISLP.load_data('Boston')
Boston.head()

,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,lstat,medv
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296,15.3,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242,17.8,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242,17.8,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222,18.7,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222,18.7,5.33,36.2


In [8]:
mu_hat = Boston['medv'].mean()
mu_hat

22.532806324110677

---

(b) Provide an estimate of the standard error of $\hat{\mu}$. Interpret this result.\
*Hint: We can compute the standard error of the sample mean by dividing the sample standard deviation by the square root of the number of observations.*

Based on the *hint*, we have:
$$\text{SEM} = \frac{\sigma}{\sqrt{n}}, \qquad\text{where } \sigma \text{ is the std of medv.}$$

In [12]:
std = Boston['medv'].std()
n = Boston.shape[0]
print(std, n, std/np.sqrt(n))

9.197104087379818 506 0.40886114749753505


Therefore, 
$$\text{SEM} = \frac{\sigma}{\sqrt{n}} = \frac{9.1941}{506} \approx 0.40886$$

---

(c) Now estimate the standard error of $\hat{\mu}$ using the bootstrap. How does this compare to your answer from (b)?

In [13]:
# use bootstrap to estimate standard error
B = 1000
boot_means = np.empty(B)
for b in range(B):
    sample = Boston.sample(n=n, replace=True)
    boot_means[b] = sample['medv'].mean()
    
boot_se = boot_means.std()
boot_se

0.40812421904775853

The **bootstrap** methods is essentially a Monte Carlo method on sampling the variance, so the calculated standard error is very close to the direct calculated in (b) within 0.001 precision.

---

(d) Based on your bootstrap estimate from (c), provide a 95% conf idence interval for the mean of `medv`. Compare it to the results obtained by using `Boston['medv'].std()` and the two standard error rule (3.9). *Hint: You can approximate a 95% confidence interval using the formula* $$[\hat{\mu} − 2\text{SE}(\hat{\mu}), \hat{\mu} + 2\text{SE}(\hat{\mu})]$$

In [14]:
print(f'[{mu_hat - 2*boot_se}, {mu_hat + 2*boot_se}]')

[21.71655788601516, 23.349054762206194]


The $95\%$ confidence interval is: $$\text{CI}_{95\%} = [21.71656, 23.34905]$$

---

(e) Based on this data set, provide an estimate, $\hat{\mu}_{med}$, for the median value of medv in the population.

In [15]:
# estimate for the median value of medv in the population
mu_med_hat = Boston['medv'].median()
mu_med_hat

21.2

$$\hat{\mu}_{med} = \text{median of MEDV} = 21.2$$

---

(f) Wenowwouldlike to estimate the standard error of $\hat{\mu}_{med}$. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [20]:
# use bootstrap to estimate standard error
B = 1000
boot_median = np.empty(B)
for b in range(B):
    sample = Boston.sample(n=n, replace=True)
    boot_median[b] = sample['medv'].median()
    
boot_se = boot_median.std()
print(f'Boostrape: {boot_se}')

Boostrape: 0.37221093146225537


$$\text{SE}_{med} = 0.3722$$
Comment: Standard error of median $<$ Standard error of mean. 

---

(g) Based on this data set, provide an estimate for the tenth percentile of `medv` in Boston census tracts. Call this quantity $\hat{\mu}_{0.1}$. (You can use the `np.percentile()` function.)

In [18]:
mu_01_hat = np.percentile(Boston['medv'], 10)
mu_01_hat

12.75

(h) Use the bootstrap to estimate the standard error of $\hat{\mu}_{0.1}$. Comment on your findings.

In [19]:
boot_percentile = np.empty(B)
for b in range(B):
    sample = Boston.sample(n=n, replace=True)
    boot_percentile[b] = np.percentile(sample['medv'], 10)

boot_se = boot_percentile.std()
boot_se

0.5193671124551495

$$\text{SE}_{0.1} = 0.5194$$

Comment: Standard error of 10th percentile $>$ Standard error of mean.